Fixed - In restricted site RSS feed is showing publicly for activity, groups, forums

jitendrabanjara1991 opened the pull request:
### All Submissions:

* [x] Have you followed the [Pull Requests Guidelines](https://github.com/buddyboss/buddyboss-platform/wiki/Submitting-Pull-Requests#pull-request-guidelines)?
* [x] Does your code follow the [WordPress' Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Fixes #2324.

### How to test the changes in this Pull Request:

1. Without login access RSS feed and API endpoints
2. Check API Link and RSS feed link here - https://docs.google.com/spreadsheets/d/1xLg4nIjDMS2DPqjNGPjPtszudosyaw4uL5yfuv_Ffbc/edit?usp=sharing (You need to change your local server URL)

### Proof Screenshots or Video

https://screencast-o-matic.com/watch/crf2jYVnzGD

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

sourabhmatolia commented:
**Important NOTE**

Need to make sure we restrict access to all third-party plugins RSS feed and Rest API End Points.

jitendrabanjara1991 commented:
@sourabhmatolia 
cc - @KartikSuthar 
I have added logic based on above requirement.
Here is the video for the proof - https://screencast-o-matic.com/watch/crh62IVflWM.
I have also check endpoint for GamiPress, LearnDash LMS, The Events Calendar. 
Please check it and let me know if remaining anyother.
Thanks.

KartikSuthar commented on src/bp-core/bp-core-catchuri.php:
```diff
@@ -1543,3 +1543,93 @@ function bp_core_change_privacy_policy_link_on_private_network( $link, $privacy_
+ * Also, remove feeds url.
+ */
+add_action('init', 'bb_removes_endpoints_and_feeds');
+function bb_removes_endpoints_and_feeds() {
+	if ( ! is_user_logged_in() ) {
+		$enable_private_network    = bp_enable_private_network();
+		$buddyboss_app_plugin_file = 'buddyboss-app/buddyboss-app.php';
+		// IF Platform active and App plugin not active.
+		if ( defined( 'BP_PLATFORM_VERSION' ) && ! is_plugin_active( $buddyboss_app_plugin_file ) ) {
+			// IF Platform 'Private Website' settings enabled.
+			// THEN Restrict all RSS and REST APIs. (reason to restrict because the site is private).
+			if ( ! $enable_private_network ) {
+				bb_restricate_rss_feed();
+				bb_restricate_rest_api();
+			}
+		}
+		// IF Platform active and App plugin also active.
+		if ( defined( 'BP_PLATFORM_VERSION' ) && is_plugin_active( $buddyboss_app_plugin_file ) ) {
+			// IF Platform 'Private Website' settings enabled and App plugin 'Private App' settings disabled.
+			// THEN Restrict all RSS but do not restrict REST APIs.
+			// (reason to restrict RSS because the site is private)
+			// (reason to not restrict rest API because a customer might want to restrict site on the web but not on App)
+			if ( ! $enable_private_network && ! bbapp_is_private_app_enabled() ) {
+				bb_restricate_rss_feed();
+			}
+			// IF Platform 'Private Website' settings enabled and App plugin 'Private App' settings also enabled.
+			// THEN Restrict all RSS and REST APIs.
+			// (reason to restrict because the site is private on both web and app)
+			if ( ! $enable_private_network && bbapp_is_private_app_enabled() ) {
+				bb_restricate_rss_feed();
+				bb_restricate_rest_api();
+			}
+		}
+	}
+}
```
@jitendrabanjara1991 can we simply update this function like this?

```suggestion
function bb_removes_endpoints_and_feeds() {
	if ( is_user_logged_in() ) {
		return;
	}

	if ( ! bp_enable_private_network() ) {
		bb_restricate_rss_feed();
	}

	if ( bp_rest_enable_private_network() ) {
		bb_restricate_rest_api();
	}
}
```

jitendrabanjara1991 replied:
@KartikSuthar 
Yes this will work. I have checked with and update it.
I have attached all condition video here.
1) platform active + app deactive
	Enable private website on platform
	Result - Restrict all RSS and REST APIs.
	working - https://screencast-o-matic.com/watch/crhX2DVfbtV
	
2) platform active + app active
	Enable private website on platform
	Disable private app on app
	Result - Restrict all RSS but do not restrict REST APIs.
	working - https://screencast-o-matic.com/watch/crhX2FVfbta

3) platform active + app active
	Enable private website on platform
	Enable private app on app
	Result - Restrict all RSS and REST APIs.
	working - https://screencast-o-matic.com/watch/crhX2YVfbue

Please check it and let me know if anything else.
Thanks

KartikSuthar commented on src/bp-core/bp-core-catchuri.php:
```diff
@@ -1545,3 +1545,77 @@ function bp_core_change_privacy_policy_link_on_private_network( $link, $privacy_
+ * @param $endpoints
+ */
+function bb_remove_all_endpoints( $endpoints ) {
+	foreach ( $endpoints as $endpoint => $details ) {
```
@jitendrabanjara1991 We need to add a filter here to exclude specific endpoints.
In BuddyBoss APP, If the app is private then we need a specific endpoint to work.

jitendrabanjara1991 replied:
@KartikSuthar 
Request changes are done. Please review it.
Thanks.

KartikSuthar commented on src/bp-core/bp-core-functions.php:
```diff
@@ -6060,4 +6060,86 @@ function bb_moderation_bg_update_moderation_data() {
+ */
+function bb_restricate_rest_api() {
+	// This will disable woocommerce endpoints.
+	if ( class_exists( 'woocommerce' ) ) {
+		remove_action( 'rest_api_init', array( WC()->api, 'register_rest_routes' ), 10 );
+	}
```
@jitendrabanjara1991 Remove this as we don't need to control or restrict other plugin API.

KartikSuthar commented on src/bp-core/bp-core-functions.php:
```diff
@@ -6060,4 +6060,86 @@ function bb_moderation_bg_update_moderation_data() {
+ *
+ * @return array $endpoints
+ */
+function bb_remove_all_endpoints( $endpoints ) {
+	if ( empty( $endpoints ) ) {
+		return;
+	}
+
+	$namespaces = rest_get_server()->get_namespaces();
+
+	// You can add exclude endpoint with an array via below filter.
+	$exclude_endpoints = apply_filters( 'bb_exclude_endpoints_from_restriction', array(
+		'/',
+		'/buddyboss/v1',
+		'/buddyboss/v1/settings',
+		'/buddyboss/v1/signup/form',
+		'/buddyboss/v1/signup/(?P<id>[\w-]+)',
+		'/buddyboss/v1/signup/activate/(?P<id>[\w-]+)',
+		'/wp/v2',
+		'/wp/v2/pages/(?P<id>[\d]+)',
+	) );
+
+	foreach ( $endpoints as $endpoint => $details ) {
+		if ( ! in_array( $endpoint, $exclude_endpoints, true ) ) {
+			unset( $endpoints[ $endpoint ] );
+		}
+	}
+
+	return $endpoints;
+}
```
@jitendrabanjara1991 Please block only the BuddyBoss REST APIs not the other one.

jitendrabanjara1991 commented:
@KartikSuthar 
As per request change, I have changed code. Now, we are blocking just our platforms API ( also exclude buddyboss-app and appboss - I have added filter if need to add then using filter can add ). We are not blocking any third party plugins API as well as default WordPress API.
So, Please review again and let me know if need any changes.
cc - @sourabhmatolia @chetansatasiya @mahedihasannoman 
Thanks.

sourabhmatolia commented:
**UPDATE**

Bumping task to next sprint because did not chance to discuss internally on how to execute this.

Based on discussion with Kartik, here's everything we might need to handle:
- Continuation from here: https://github.com/buddyboss/buddyboss-platform/pull/2331#issuecomment-836191757
- Need to restrict all possible REST APIs like from WP and Third Party plugins also.
- Need to restrict all possible RSS Feed like from WP and Third Party plugins also.
- Need to provide option to exclude specific REST API and specific RSS Feed to make public.

sourabhmatolia commented:
@jitendrabanjara1991 @KartikSuthar 

**Quick R & D action points for this task to proceed further:**

- We will provide settings to set certain RSS feeds or Rest Endpoints public.
See how we can make a specific set of posts, pages, or any other content public using just a URL in the settings.

- Do R & D on how we can allow REST API access to just BuddyBoss app but not otherwise?

- Related to this task: https://github.com/buddyboss/buddyboss-app/issues/1996
Check how we can allow access to authenticate users using this endpoint "/wp-json/wp/v2/users"
But we do not return a response with data?

KartikSuthar commented:
> @jitendrabanjara1991 @KartikSuthar
> 
> **Quick R & D action points for this task to proceed further:**
> 
> * We will provide settings to set certain RSS feeds or Rest Endpoints public.
>   See how we can make a specific set of posts, pages, or any other content public using just a URL in the settings.
>> We can do this.
> * Do R & D on how we can allow REST API access to just BuddyBoss app but not otherwise?
>> Need to check with App Plugin developer.
> * Related to this task: https://github.com/buddyboss/buddyboss-app/issues/1996
>   Check how we can allow access to authenticate users using this endpoint "/wp-json/wp/v2/users"
>   But we do not return a response with data?
>> Need to check with App developer.
> 
> Once we have answers to the questions above, we will proceed with the final implementation here. Thanks.

singhleo commented:
```
Related to this task: https://github.com/buddyboss/buddyboss-app/issues/1996
Check how we can allow access to authenticate users using this endpoint "/wp-json/wp/v2/users"
But we do not return a response with data?
```

Users can authenticate with BuddyBoss App Authenticate Endpoints without requiring of users endpoint. 
But for further logged-in operations, we need users' endpoints .. so we cannot block users' endpoints for authenticated users.

JetalDobariya commented:
Hello, @sourabhmatolia  @firozwp 

We need to bump this PR in up-coming sprint reasons of for this PR I need to install npm and node JS and also need to do few setups with App plugin using shell commands , I already did this process with @Jitendra Banjara and @Sumit  but facing something issues , we didn't catch up what is an error exactly, So I need to check with @Kartik Suthar.
One more thing This PR is too large PR so QA-team will need enough time for testing. 

Thanks!
CC : @sumitbuddyboss

JetalDobariya commented:
Hello @jitendrabanjara1991 

I tested this PR in qa-ongoing branch, RSS Feed end points are working fine.
I attached screenshots in the sheet tab name "RSS Feed"
https://docs.google.com/spreadsheets/d/1DHmg07d95x_cPVtDu34EFXPSqrFbI8XH/edit#gid=973374751

Thanks!

jitendrabanjara1991 commented:
@JetalDobariya 
@asaquzzamanbb has fixed issue related to particular user's API and now its working fine. So, Please check it. Apart from that, We checked API's endpoint which list outed not working in spreadsheet, They all are working fine. We checked with @JetalDobariya on call.
@JetalDobariya please verify it.
cc - @harshalQA 
Thanks.

JetalDobariya commented:
Hello @jitendrabanjara1991

I verified this issue again and working fine. See videos proofs.
https://www.loom.com/share/08c249dd392b40758869cf93b18f7088

https://docs.google.com/spreadsheets/d/1DHmg07d95x_cPVtDu34EFXPSqrFbI8XH/edit#gid=840517568

Thanks!
CC : @harshalQA

jitendrabanjara1991 commented:
@JetalDobariya 
For the RSS feed restriction - Can you please quick recheck because I have added one minor condition for feed because its creating issue on the app side.
Thanks.

JetalDobariya commented:
Hello @jitendrabanjara1991 

I did verify again all the Private RSS Feeds End points and all the EndPoints are working perfectly with Login and logout user.
Also i verify with create Post and Page Slug with "feed" and it is working fine.
See in below video  :
https://www.loom.com/share/8541ec7f29ed43c6b868b957e36db397

Thanks!

jitendrabanjara1991 commented:
Code Status
Merged in qa-approved
Thanks

jitendrabanjara1991 commented:
Code Status
Merged in qa-approved.
Thanks

JetalDobariya commented:
Hello @jitendrabanjara1991 

I did verify all the RSS feed end points in qa-approved branch, now all the end points are working fine with Login and without login.
Also I checked create Post with "feed" word and checked slug as well as it is working fine.

Thanks!
CC : @sourabhmatolia

JetalDobariya commented:
Hello @jitendrabanjara1991 

I tested above given issue in qa-approved branch, now all the issues are getting resolved.
All API endpoitns are working as per App, Platform and Rest API's settings.
Here , I attached video Please check.
https://loom.com/share/277bf4459b9942edabfc39008c870b57
In that I facing issues in three API which is listed below :
If Private Website and Rest API both Enabled then getting error while hit below API's.
wp-json/wp/v2/plugins --> https://prnt.sc/220o4ow
wp-json/wp/v2/themes --> https://prnt.sc/220nz3y
wp-json/wp/v2/settings  --> https://prnt.sc/220nrh5

-> If Private Website and Rest API both are enabled then Sign-up form is working.
-> If Buddyboss App is Dactive then Rest API text area will be hide.
https://loom.com/share/6035caee17f84957b384ab60be9cffd2

Please check and let me know still you facing same issue.
Thanks!
CC : @sourabhmatolia @KartikSuthar

jitendrabanjara1991 commented:
> Hello @jitendrabanjara1991
> 
> I tested above given issue in qa-approved branch, now all the issues are getting resolved. All API endpoitns are working as per App, Platform and Rest API's settings. Here , I attached video Please check. https://loom.com/share/277bf4459b9942edabfc39008c870b57 In that I facing issues in three API which is listed below : If Private Website and Rest API both Enabled then getting error while hit below API's. wp-json/wp/v2/plugins --> https://prnt.sc/220o4ow wp-json/wp/v2/themes --> https://prnt.sc/220nz3y wp-json/wp/v2/settings --> https://prnt.sc/220nrh5
> 
> -> If Private Website and Rest API both are enabled then Sign-up form is working. -> If Buddyboss App is Dactive then Rest API text area will be hide. https://loom.com/share/6035caee17f84957b384ab60be9cffd2
> 
> Please check and let me know still you facing same issue. Thanks! CC : @sourabhmatolia @KartikSuthar

@JetalDobariya 
wp-json/wp/v2/plugins --> https://prnt.sc/220o4ow
wp-json/wp/v2/themes --> https://prnt.sc/220nz3y
wp-json/wp/v2/settings --> https://prnt.sc/220nrh5
Above three API endpoints, We can't allow them because its restricted from WP.

Thanks.
@sourabhmatolia @firozwp @KartikSuthar

sourabhmatolia commented:
**UPDATE**

Based on discussion with @JetalDobariya, rejecting task from this release.
We still have many blocker issues here which I can find very easily in my review.

I will process all the issues found, what all to check moving forward, how to check, and update here. Thanks.

sourabhmatolia commented:
@jitendrabanjara1991 @JetalDobariya @KartikSuthar 

Following up here

**Here's the list of outstanding issues from my review:**

- Number 3 issues shared here still doesn't work:
https://github.com/buddyboss/buddyboss-platform/pull/2331#issuecomment-986528985

- Number 1 resolved that means REST APIs accessible when "Private website" is enabled.
But again when REST API restricted as well and if we exclude "buddyboss/v1/activity" or "wp/v2/pages" or "buddyboss/v1/groups" APIs  then it doesn't make that accessible to public.

- Reported this issue but did not cross check again if it works fine now:
Enable App Plugin > Enable Private app > REST API restricted now.
Exclude: "buddyboss/v1/members" or "buddyboss/v1/activity"
Still API doesn't work.
Doesn't work with WP API also:
"wp/v2/pages"

- App Plugin active > Private App not selected > Not getting response for this:
wp-json/buddyboss-app/v1/signup

- Restricted RSS Feed > excluded "/hello-world/feed" but still this doesn't work:
domain.com/hello-world/feed
Same excluded "/feed" but still this doesn't work:
domain.com/feed

- App Plugin active > Private App not enabled > "Private REST APIs" checkbox shows disabled but we should still show "BuddyBoss App is enabled" notice no matter if app plugin "Private App" option enabled or no.

- Question: I still don't see document showing list of APIs enabled for public for non logged in always:
https://prnt.sc/22anufw

---

DEV and QA Moving forward please check everything thoroughly. We will save lot of time

I am simply trying APIs from here and finding issues:
RSS Feeds: https://wordpress.org/support/article/wordpress-feeds/
WordPress REST APIs: https://developer.wordpress.org/rest-api/
Platform REST APIs: https://www.buddyboss.com/resources/api/
App REST APIs: https://www.buddyboss.com/resources/api/app/

Firstly I tried to see if all these are open when REST APIs not restricted.
Next I tried to restrict and exclude specific to check excluded one works fine or no.
Hoping we make good progress here and try to get it all resolves ASAP. Thanks.

jitendrabanjara1991 commented:
Hey @sourabhmatolia 

Please check below updates.

Number 3 issues shared here still doesn't work:
Fixed - In restricted site RSS feed is showing publicly for activity, groups, forums #2331 (comment)

Number 1 resolved that means REST APIs accessible when "Private website" is enabled.
But again when REST API restricted as well and if we exclude "buddyboss/v1/activity" or "wp/v2/pages" or "buddyboss/v1/groups" APIs then it doesn't make that accessible to public.

Reported this issue but did not cross check again if it works fine now:
Enable App Plugin > Enable Private app > REST API restricted now.
Exclude: "buddyboss/v1/members" or "buddyboss/v1/activity"
Still API doesn't work.
Doesn't work with WP API also:
"wp/v2/pages"

-> **Above issue has been solved. Here is the video as the proof - https://screencast-o-matic.com/watch/crlYoPV2KOp**

App Plugin active > Private App not selected > Not getting response for this:
wp-json/buddyboss-app/v1/signup
-> **As I know we get signup data here. So this API call after login so that's why its require logging. For more info I will cross verify and also discuss with @parmarhardip @KartikSuthar and @shakilAshraful for the same.**

App Plugin active > Private App not enabled > "Private REST APIs" checkbox shows disabled but we should still show "BuddyBoss App is enabled" notice no matter if app plugin "Private App" option enabled or no.
-> **Done - https://screencast-o-matic.com/watch/crlYDuV2Kw2**

Restricted RSS Feed > excluded "/hello-world/feed" but still this doesn't work:
domain.com/hello-world/feed
Same excluded "/feed" but still this doesn't work:
domain.com/feed
-> **Done - https://screencast-o-matic.com/watch/crlYDzV2Kxk**

Question: I still don't see document showing list of APIs enabled for public for non logged in always:
-> Here is the list of APIs which we are excluding API as default.
     Below APIs endpoint are excluding from platform side.
        1.  /buddyboss/v1/signup/form
        2.  /buddyboss/v1/signup/(?P<id>[\w-]+)
        3.  /buddyboss/v1/signup/activate/(?P<id>[\w-]+)
        4.  /buddyboss/v1/settings
        5.  /buddyboss/v1/signup

     Below APIs endpoint are excluding from App plugin side.
        1.  /buddyboss-app/core/v1/settings/(?P<app_id>[a-zA-Z0-9-]+)
        2.  /buddyboss-app/core/v1/settings
        3.  /buddyboss/v1/settings
        4.  /buddyboss-app/core/v1/app_details
        5.  /buddyboss-app/core/v1/menu/(?P<app_id>[a-zA-Z0-9-]+)
        6.  /buddyboss-app/v1/signup/form
        7.  /buddyboss/v1/signup/form
        8.  /buddyboss-app/v1/signup/verify
        9.  /buddyboss-app/auth/v1/verify
        10.  /buddyboss-app/v1/signup
        11.  /buddyboss/v1/signup
        12.  /buddyboss-app/auth/v1/jwt/token
        13.  /buddyboss-app/auth/v2/jwt/login
        14.  /buddyboss-app/core/v1/device_token
        15.  /buddyboss-app/auth/v1/forgot
        16.  /buddyboss-app/iap/v1/products
        17.  /buddyboss-app/iap/v1/order
        18.  /buddyboss-app/iap/v2/order

      @asaquzzamanbb and @shakilAshraful correct me here if I missed any endpoint here which we are keeping publicly for non-logged in user.
       @sourabhmatolia I am not aware about document like where we need to add above all endpoints as doc?

Thanks.

asaquzzamanbb commented:
> Hey @sourabhmatolia
> 
> Please check below updates.
> 
> Number 3 issues shared here still doesn't work: Fixed - In restricted site RSS feed is showing publicly for activity, groups, forums #2331 (comment)
> 
> Number 1 resolved that means REST APIs accessible when "Private website" is enabled. But again when REST API restricted as well and if we exclude "buddyboss/v1/activity" or "wp/v2/pages" or "buddyboss/v1/groups" APIs then it doesn't make that accessible to public.
> 
> Reported this issue but did not cross check again if it works fine now: Enable App Plugin > Enable Private app > REST API restricted now. Exclude: "buddyboss/v1/members" or "buddyboss/v1/activity" Still API doesn't work. Doesn't work with WP API also: "wp/v2/pages"
> 
> -> **Above issue has been solved. Here is the video as the proof - https://screencast-o-matic.com/watch/crlYoPV2KOp**
> 
> App Plugin active > Private App not selected > Not getting response for this: wp-json/buddyboss-app/v1/signup -> **As I know we get signup data here. So this API call after login so that's why its require logging. For more info I will cross verify and also discuss with @parmarhardip @KartikSuthar and @shakilAshraful for the same.**
> 
> App Plugin active > Private App not enabled > "Private REST APIs" checkbox shows disabled but we should still show "BuddyBoss App is enabled" notice no matter if app plugin "Private App" option enabled or no. -> **Done - https://screencast-o-matic.com/watch/crlYDuV2Kw2**
> 
> Restricted RSS Feed > excluded "/hello-world/feed" but still this doesn't work: domain.com/hello-world/feed Same excluded "/feed" but still this doesn't work: domain.com/feed -> **Done - https://screencast-o-matic.com/watch/crlYDzV2Kxk**
> 
> Question: I still don't see document showing list of APIs enabled for public for non logged in always: -> Here is the list of APIs which we are excluding API as default.
> 
> ```
>  Below APIs endpoint are excluding from platform side.
>     1.  /buddyboss/v1/signup/form
>     2.  /buddyboss/v1/signup/(?P<id>[\w-]+)
>     3.  /buddyboss/v1/signup/activate/(?P<id>[\w-]+)
>     4.  /buddyboss/v1/settings
>     5.  /buddyboss/v1/signup
>  
>  Below APIs endpoint are excluding from App plugin side.
>     1.  /buddyboss-app/core/v1/settings/(?P<app_id>[a-zA-Z0-9-]+)
>     2.  /buddyboss-app/core/v1/settings
>     3.  /buddyboss/v1/settings
>     4.  /buddyboss-app/core/v1/app_details
>     5.  /buddyboss-app/core/v1/menu/(?P<app_id>[a-zA-Z0-9-]+)
>     6.  /buddyboss-app/v1/signup/form
>     7.  /buddyboss/v1/signup/form
>     8.  /buddyboss-app/v1/signup/verify
>     9.  /buddyboss-app/auth/v1/verify
>     10.  /buddyboss-app/v1/signup
>     11.  /buddyboss/v1/signup
>     12.  /buddyboss-app/auth/v1/jwt/token
>     13.  /buddyboss-app/auth/v2/jwt/login
>     14.  /buddyboss-app/core/v1/device_token
>     15.  /buddyboss-app/auth/v1/forgot
>     16.  /buddyboss-app/iap/v1/products
>     17.  /buddyboss-app/iap/v1/order
>     18.  /buddyboss-app/iap/v2/order
> ```
> 
> @asaquzzamanbb and @shakilAshraful correct me here if I missed any endpoint here which we are keeping publicly for non-logged in user. @sourabhmatolia I am not aware about document like where we need to add above all endpoints as doc?
> 
> Thanks.

@jitendrabanjara1991 Looks good.

sumitbuddyboss commented:
Hello @jitendrabanjara1991 

**Question:**
If my BB app plugin is activated and "Private App" is deactivated then I can not enable "Private REST APIs" its checkbox is disabled, if users want to restrict REST Apis then how can do that?

**Video:**
https://www.loom.com/share/beb59e47d8334d8cb29b7580ec311877

cc @sourabhmatolia

jitendrabanjara1991 commented:
Hello @sumitbuddyboss 
If BB app plugin is activated and user want to restrict REST APIs then require to enable Private App options.
@sourabhmatolia correct me here if anything wrong.
Thanks

sourabhmatolia commented:
@sumitbuddyboss @jitendrabanjara1991 

>If my BB app plugin is activated and "Private App" is deactivated then I can not enable "Private REST APIs" its checkbox is disabled, if users want to restrict REST Apis then how can do that?

IF App plugin is active but not private then we cannot restrict APIs because all these APIs are being used by App Native for non-logged-in view in the app i.e. "Continue without logging in".
That's why the Platform restrict API option is disabled. Thanks

sumitbuddyboss commented:
Hello @jitendrabanjara1991

**I notice one issue:**
When the **Private RSS Feeds** is enabled then the feed should not allow without login but I can access the below feed without login user.
http://platformtesting.com/?s=test&feed=rss2

**Issue video:**
https://www.loom.com/share/142268d4128345a6969b5f5c879c33af

Thank you

sumitbuddyboss commented:
Hello @sourabhmatolia 

**Query:**
A "Private website" is enabled and "Private RSS Feeds"	is disabled then we can access the feed with login and logged-out users, Do we need to restrict the feed when the private website is enabled and "Private RSS Feeds" is disabled?

**Video:**
https://www.loom.com/share/0fb5380f193e49c1841bfda4e7df8224

cc @jitendrabanjara1991 

Thank you

jitendrabanjara1991 commented:
> Hello @jitendrabanjara1991
> 
> **I notice one issue:** When the **Private RSS Feeds** is enabled then the feed should not allow without login but I can access the below feed without login user. http://platformtesting.com/?s=test&feed=rss2
> 
> **Issue video:** https://www.loom.com/share/142268d4128345a6969b5f5c879c33af
> 
> Thank you

@sourabhmatolia 

If someone change permalink as **Plain** permalink ( https://prnt.sc/24irr82 ) then feed URL will generate as per below ( https://prnt.sc/24irsyp )
- https://localhost/projects/testing-app/?feed=rss2
- https://localhost/projects/testing-app/?feed=comments-rss2

If we set permalink as **Post Name** then URL will generate as per below ( https://prnt.sc/24ispaw )
- https://localhost/projects/testing-app/feed/
- https://localhost/projects/testing-app/comments/feed/

Right now we are allowing to exclude as per this example **e.g. /post-name/feed/** which will not work if permalink set as **Plain**.

So can we allow to enter below type of examples in Public Content for RSS Feed.
- feed=rss2 ( If plain permalink then we can use this )
- feed=comments-rss2 ( If plain permalink then we can use this )
- /feed/ ( If any other permalink then we can use this )
- /news/feed/ ( If any other permalink then we can use this )

Here is the video - https://screencast-o-matic.com/watch/crlTbyV2Ru1

Let me know your thought here.
cc - @sumitbuddyboss @KartikSuthar @firozwp 
Thanks.

sourabhmatolia commented:
@sumitbuddyboss 

>Query:
A "Private website" is enabled and "Private RSS Feeds" is disabled then we can access the feed with login and logged-out users, Do we need to restrict the feed when the private website is enabled and "Private RSS Feeds" is disabled?
Video:
https://www.loom.com/share/0fb5380f193e49c1841bfda4e7df8224

No we don't want to restrict RSS Feed when "Private website" is enabled.
That's the reason we provided a setting to manually disable RSS feed using the new option provided with this task.

---

@jitendrabanjara1991 

>Responding your concern shared above here:
https://github.com/buddyboss/buddyboss-platform/pull/2331#issuecomment-1001050600

When we set permalink to Plain then no profile URL works in frontend. There are a bunch of stuff breaks there which we are planning to handle in the future.
I think anything other than "Post name" or "Custom" doesn't work properly with the platform which we will need to handle later.
In the future, we might force customers to use the "Post name" permalink structure or something like that.
So I don't think we should bother about this at the moment. we can come back. 

Thanks.

sumitbuddyboss commented:
Hello @jitendrabanjara1991 

I've cross-checked all the endpoints based on our new settings and everything is working now as expected.

If Private Website Enable > Check Endpoints - **PASS**
https://www.loom.com/share/13cde66854b04f0c8e37ec3f7a0cdd3a

If Private REST APIs	 Enable and Buddyboss App is disabled - **PASS**
https://www.loom.com/share/c08baf83637b4c718c018f2ad86c6411

If Private app(buddyboss app) Enable > Private REST APIs enable - **PASS**
https://www.loom.com/share/e6113f927f2a4aecb465add22a3a1dcd
https://www.loom.com/share/679609f50c854f6b808a4bf917d73490

If Buddboss app is disable > Private Website enable > Private REST APIs - **PASS**
https://www.loom.com/share/74d081ee67ee48a497066102c24305d1

When REST API is restricted then "/wp-json/wp/v2/users" should not return user details and at the same time, it should work fine in App Native because app native used this endpoint for authentication. - **PASS**
https://prnt.sc/24qujap

Private RSS Feeds > When private RSS enabled and disabled, also exclude feed - **PASS** 
https://www.loom.com/share/c67d2588bed64969a082217e438c9812

**Default Endpoints**
Below endpoint needs user login. 
/buddyboss/v1/signup/(?P<id>[\w-]+) - It's return user detail https://prnt.sc/24rb2qn
/buddyboss/v1/signup/activate/(?P<id>[\w-]+) - It's not a GET method it's PATCH method https://prnt.sc/24rbcp4
/buddyboss-app/iap/v1/order
/buddyboss-app/iap/v2/order

**List of all the APIs endpoints**
https://docs.google.com/spreadsheets/d/1rmz9EbcmU1ZcHeHQ8bwgFNnIU1xzbaGbFdR0UJSdUDk/edit#gid=0

Thank you.

jitendrabanjara1991 commented:
Code status
- Merged in qa-approved
Thanks.

sourabhmatolia commented:
@jitendrabanjara1991 @sumitbuddyboss 

Checked functionality again and I am finding issues with core functionalities still.

Check Video:
https://www.loom.com/share/4375aea75cfa45adb292ed48fc907659
(app plugin was not active when I ran the following test cases)

- I have just disabled RSS feed access but now I cannot access the homepage or anything else anymore.

- Restrict RSS feed only
Excluded "/comments/"
This URL still not accessible.
http://localhost/buddyboss-platform/comments/feed/
  - Same for "/hello-world/"
http://localhost/buddyboss-platform/hello-world/feed
  - Same for "/uncategorized/"
http://localhost/buddyboss-platform/category/uncategorized/feed/

- Restricted REST API only.
Excluded "wp/v2/pages/4330"
But this doesn't work:
{{platform-domain}}/wp-json/wp/v2/pages/4330
  - Same issue, exlcuded "wp/v2/users/2"
But this doesn't work:
{{platform-domain}}/wp-json/wp/v2/users/2
  - Same issue, excluded "wp/v2/taxonomies"
  - Same issue, excluded "buddyboss/v1/activity"

jitendrabanjara1991 commented:
@sourabhmatolia 
I have checked your video and look like something missed. 
Like For API - If you want to exclude any API then below type will work as per instruction ( https://prnt.sc/25pg7yg )

- wp-json/wp/v2/posts
- wp-json/wp/v2/users
- wp-json/wp/v2/pages/249
- wp-json/buddyboss/v1/activity

I have tested it and its working fine. Please check video - https://screencast-o-matic.com/watch/c3Vn28Voj2g
For RSS feed - If you want to exclude any feed then below type will work as per instruction ( https://prnt.sc/25pi5m1 )

- /comments/feed/
- /hello-world/feed/
- /uncategorized/feed/

Its also working fine if we add feed URL like above.

I have just disabled RSS feed access but now I cannot access the homepage or anything else anymore. - That was the issue. The issue was happening because feed action was calling on init hook that's why. I fixed it - https://screencast-o-matic.com/watch/c3Vno0Vojod

@sumitbuddyboss Please verify it.

Thanks.

sourabhmatolia commented:
@jitendrabanjara1991 

>Regarding "wp-json/buddyboss/v1/activity" vs "buddyboss/v1/activity" for REST API
OR
"/comments/" vs "/comments/feed/" for RSS Feed URL

Logic explained above is not accurate.  Please check Video: 
https://www.loom.com/share/ef882c4877ef47ffafd0194c420f4b64

jitendrabanjara1991 commented:
> @jitendrabanjara1991
> 
> > Regarding "wp-json/buddyboss/v1/activity" vs "buddyboss/v1/activity" for REST API
> > OR
> > "/comments/" vs "/comments/feed/" for RSS Feed URL
> 
> Logic explained above is not accurate. Please check Video: https://www.loom.com/share/ef882c4877ef47ffafd0194c420f4b64
> 
> When I reported the issues above then also I was using a fragment of the URL: https://prnt.sc/25rxlui Not sure why we didn't bring this then.

@sourabhmatolia 
We allow fragment and tested it.
For RSS Feed - https://screencast-o-matic.com/watch/c3VeICVoQsM
For Rest API - https://screencast-o-matic.com/watch/c3VeIMVoQL3
Thanks.

sumitbuddyboss commented:
Hello @jitendrabanjara1991 

We have checked the URL and URL fragment and now it's working as expected.

I've checked the below cases, Please find the below videos.

1) If Private Website enables - PASS
https://www.loom.com/share/365bde5421554faaa0a129a1b028812c

2) Private REST APIs - PASS
https://www.loom.com/share/75f8042ab0d343d28f5391c6ce01c843

3) Feed - PASS
https://www.loom.com/share/b75734695cc940bb9c3418b02225ff0b

Feed -Cross check with other settings
https://www.loom.com/share/fc65d0c21a144e47a4732dc0e76debad

4) All the settings are disabled and the app is enabled - PASS
https://www.loom.com/share/da9c7267faf84ea69e13db0af64b631f

5) App is enabled and cross-check with the other settings - PASS
https://www.loom.com/share/f184d11ef4524a1cb472267e387103bd

Thank you

KartikSuthar commented on src/bp-core/admin/bp-core-admin-settings.php:
```diff
@@ -1592,3 +1592,95 @@ function bb_after_update_activity_settings( $tab_name ) {
+	$checked_checkbox = bp_enable_private_rest_apis();
+	if ( function_exists( 'bbapp_is_private_app_enabled' ) ) {
+		if ( true === bbapp_is_private_app_enabled() ) {
+			$disable_field    = false;
```
```suggestion
			$disable_field = false;
```

KartikSuthar commented on src/bp-core/admin/bp-core-admin-settings.php:
```diff
@@ -1592,3 +1592,95 @@ function bb_after_update_activity_settings( $tab_name ) {
+	<input id="bb-enable-private-rest-apis" name="bb-enable-private-rest-apis" type="checkbox" value="1"<?php checked( $checked_checkbox ); ?><?php disabled( $disable_field ); ?>/>
+	<label for="bb-enable-private-rest-apis"><?php esc_html_e( 'Restrict REST APIs access to only logged-in members', 'buddyboss' ); ?></label>
+	<?php
+	printf(
+		'<p class="description">%s</p>',
+		sprintf(
+			__( 'Login and <a href="%s">Registration</a> APIs will remain publicly visible.', 'buddyboss' ),
+			add_query_arg(
+				array(
+					'page' => 'bp-pages',
+				),
+				admin_url( 'admin.php' )
+			)
+		)
+	);
```
```suggestion
	<p class="description">
		<?php
		printf(
			wp_kses_post(
				/* translators: Registration link. */
				__( 'Login and %s APIs will remain publicly visible.', 'buddyboss' )
			),
			sprintf(
				'<a href="%s">' . esc_html__( 'Registration', 'buddyboss' ) . '</a>',
				esc_url(
					add_query_arg(
						array( 'page' => 'bp-pages' ),
						admin_url( 'admin.php' )
					)
				)
			)
		);
		?>
	</p>
```

KartikSuthar commented on src/bp-core/admin/bp-core-admin-settings.php:
```diff
@@ -1592,3 +1592,95 @@ function bb_after_update_activity_settings( $tab_name ) {
+			)
+		)
+	);
+	if ( function_exists( 'bbapp_is_private_app_enabled' ) && false === bbapp_is_private_app_enabled() ) {
+		printf(
+			'<div class="bp-feedback info"><span class="bp-icon" aria-hidden="true"></span><p>%s</p></div>',
+			sprintf(
+				__( 'BuddyBoss App plugin is active. Please go to <a href="%s">settings</a> to enable "Private App" that will allow you to enable or disable REST APIs access for non-logged-in members.', 'buddyboss' ),
+				add_query_arg(
+					array(
+						'page' => 'bbapp-settings',
+					),
+					admin_url( 'admin.php' )
+				)
+			)
+		);
+	}
```
```suggestion
	if ( function_exists( 'bbapp_is_private_app_enabled' ) && false === bbapp_is_private_app_enabled() ) {
		?>
		<div class="bp-feedback info">
			<span class="bp-icon" aria-hidden="true"></span>
			<p>
				<?php
				printf(
					wp_kses_post(
					/* translators: Settings link. */
						__( 'BuddyBoss App plugin is active. Please go to %s to enable "Private App" that will allow you to enable or disable REST APIs access for non-logged-in members.', 'buddyboss' )
					),
					sprintf(
						'<a href="%s">' . esc_html__( 'settings', 'buddyboss' ) . '</a>',
						esc_url(
							add_query_arg(
								array( 'page' => 'bbapp-settings' ),
								admin_url( 'admin.php' )
							)
						)
					)
				);
				?>
			</p>
		</div>
		<?php
	}
```

KartikSuthar commented on src/bp-core/bp-core-options.php:
```diff
@@ -1806,17 +1806,32 @@ function bp_core_display_name_format( $default = 'first_name' ) {
-			)
-		)
-	);
+	// If BB APP active + BB APP Private + BB Private REST APIs enable then and then all rest will be private.
+	if (
+		function_exists( 'bbapp_is_private_app_enabled' ) // buddyboss-app is active.
+		&& true == (bool) bbapp_is_private_app_enabled() // private app is enabled.
+		&& true == (bool) bp_enable_private_rest_apis() // BB private rest api is enabled.
+	) {
+		$retval = (bool) bbapp_is_private_app_enabled();
+
+		// If BB APP active + private API is disable.
+	} elseif (
+		function_exists( 'bbapp_is_private_app_enabled' ) // buddyboss-app is active.
+		&& false == (bool) bbapp_is_private_app_enabled() // private app is disable.
+	) {
+		$retval = (bool) bbapp_is_private_app_enabled();
+
+		// If platform rest API private.
+	} else {
+		$retval = (bool) bp_enable_private_rest_apis();
+	}
+
```
```suggestion
	$retval = false;

	if (
		true === (bool) bp_enable_private_rest_apis() &&
		(
			(
				function_exists( 'bbapp_is_private_app_enabled' ) // buddyboss-app is active.
				&& true === (bool) bbapp_is_private_app_enabled() // private app is disable.
			)
			|| ! function_exists( 'bbapp_is_private_app_enabled' )
		)
	) {
		$retval = true;
	}

```
